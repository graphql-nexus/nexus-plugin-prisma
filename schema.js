module.exports = require('./dist/schema-entrypoint')
