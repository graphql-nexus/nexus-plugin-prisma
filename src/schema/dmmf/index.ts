export * from './utils'
export * from './transformer'
export * from './DmmfDocument'
export * from './DmmfTypes'
