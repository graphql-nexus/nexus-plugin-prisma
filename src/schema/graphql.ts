//
// GraphQL root types data & helpers
//

export type RootName = 'Query' | 'Mutation' | 'Subscription'

export const rootNames = {
  Query: 'Query',
  Mutation: 'Mutation',
  Subscription: 'Subscription',
} as const

export const rootNameValues: RootName[] = Object.values(rootNames)

export const isRootName = (x: any): x is RootName => rootNameValues.includes(x)

//
// GraphQL scalar types data & helpers
//

export type ScalarName = 'Int' | 'Float' | 'String' | 'ID' | 'Boolean'

export const scalarsNameValues: ScalarName[] = ['Int', 'Float', 'String', 'ID', 'Boolean']

export const isScalarType = (name: string): boolean => scalarsNameValues.includes(name as any)
