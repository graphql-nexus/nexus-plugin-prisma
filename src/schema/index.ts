import { plugin } from '@nexus/schema'
import { build as buildNexusPrismaTypes, Options, InternalPublicOptions } from './builder'
import { CRUD_NOT_ENABLED } from './warnings'

/**
 * Create a nexus-plugin-prisma instance to be passed into the Nexus plugins array.
 *
 * The nexus plugin extends the Nexus DSL with `t.model` and `t.crud`. These
 * allow you to expose data from your Prisma layer and operations against that
 * data. Once this plugin is installed, you can do things such as:
 *
 * ```ts
 * objectType({
 *   name: 'User',
 *   definition(t) {
 *     t.model.id()
 *     t.model.email()
 *   }
 * })
 *
 * queryType({
 *   definition (t) {
 *      t.crud.user()
 *      t.crud.users({ filtering: true, ordering: true })
 *    }
 * })
 * ```
 *
 * You must ensure the Prisma Client JS has been generated prior as it provides a
 * data representation of the available models and CRUD operations against them.
 *
 * Like Nexus, NexusPrisma has its own Typegen. By default It will be generated
 * synchronously when Nexus runs the plugin, _if_ `process.env.NODE_ENV` is
 * `undefined` or `"development"`. Typegen can be explicitly enabled or disabled
 * via the shouldGenerateArtifacts option. This mirrors Nexus' own typegen
 * approach. This system will change once the Nexus plugin system has first
 * class support for typegen.
 */
export function nexusSchemaPrisma(options?: Options) {
  const allOptions: InternalPublicOptions = options ?? {}
  let wasCrudUsedButDisabled: null | (() => boolean) = null

  return plugin({
    name: 'nexus-plugin-prisma',
    onInstall: (nexusBuilder) => {
      const { types, wasCrudUsedButDisabled: wasCrudUsed } = buildNexusPrismaTypes({
        ...allOptions,
        nexusBuilder,
      })

      wasCrudUsedButDisabled = wasCrudUsed

      return {
        types,
      }
    },
    onBeforeBuild() {
      if (wasCrudUsedButDisabled?.() === true) {
        if (allOptions.initializedByFramework === true) {
          CRUD_NOT_ENABLED.framework()
        } else {
          CRUD_NOT_ENABLED.schema()
        }
      }
    },
  })
}

export { Options } from './builder'
