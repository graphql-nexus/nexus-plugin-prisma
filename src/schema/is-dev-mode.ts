export function isDevMode() {
  return Boolean(!process.env.NODE_ENV || process.env.NODE_ENV === 'development')
}
