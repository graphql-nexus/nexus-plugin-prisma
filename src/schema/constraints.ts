import { DmmfDocument } from './dmmf'

/**
 * Find the unique identifiers necessary to indentify a field
 *
 * Unique fields for a model can be one of (in this order):
 * 1. One (and only one) field with an @id annotation
 * 2. Multiple fields with @@id clause
 * 3. One (and only one) field with a @unique annotation (if there are multiple, use the first one)
 * 4. Multiple fields with a @@unique clause
 */
export function resolveUniqueIdentifiers(typeName: string, dmmf: DmmfDocument): string[] {
  const model = dmmf.getModelOrThrow(typeName)

  // Try finding 1.
  const singleIdField = model.fields.find((f) => f.isId)

  if (singleIdField) {
    return [singleIdField.name]
  }

  // Try finding 2.
  if (model.idFields && model.idFields.length > 0) {
    return model.idFields
  }

  const singleUniqueField = model.fields.find((f) => f.isUnique)

  if (singleUniqueField) {
    return [singleUniqueField.name]
  }

  if (model.uniqueFields && model.uniqueFields.length > 0) {
    return model.uniqueFields[0]
  }

  throw new Error(`Unable to resolve a unique identifier for the Prisma model: ${model.name}`)
}

export function findMissingUniqueIdentifiers(
  data: Record<string, any>,
  uniqueIdentifiers: string[]
): string[] | null {
  const missingIdentifiers: string[] = []

  for (const identifier of uniqueIdentifiers) {
    if (!data[identifier]) {
      missingIdentifiers.push(identifier)
    }
  }

  if (missingIdentifiers.length > 0) {
    return missingIdentifiers
  }

  return null
}

export function buildWhereUniqueInput(data: Record<string, any>, uniqueIdentifiers: string[]) {
  if (uniqueIdentifiers.length === 1) {
    return pickFromRecord(data, uniqueIdentifiers)
  }

  const compoundName = uniqueIdentifiers.join('_')

  return {
    [compoundName]: pickFromRecord(data, uniqueIdentifiers),
  }
}

function pickFromRecord(record: Record<string, any>, keys: string[]) {
  const output: Record<string, any> = {}

  for (const identifier of keys) {
    if (record[identifier]) {
      output[identifier] = record[identifier]
    }
  }

  return output
}
