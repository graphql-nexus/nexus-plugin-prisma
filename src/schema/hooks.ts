import { colors } from './colors'
import { isDevMode } from './is-dev-mode'

export type OnUnknownFieldName = (info: {
  unknownFieldName: string
  error: Error
  validFieldNames: string[]
  typeName: string
}) => void

export type OnUnknownFieldType = (info: {
  unknownFieldType: string
  error: Error
  typeName: string
  fieldName: string
}) => void

export type OnUnknownArgName = (info: {
  unknownArgNames: string[]
  error: Error
  typeName: string
  fieldName: string
}) => void

export type OnUnknownPrismaModelName = (info: { unknownPrismaModelName: string; error: Error }) => void

type Hooks = OnUnknownFieldType | OnUnknownFieldName | OnUnknownArgName | OnUnknownPrismaModelName | undefined

export function raiseErrorOrTriggerHook<T extends Hooks>(
  hook: T,
  params: Parameters<Exclude<T, undefined>>[0],
  message: string,
  stage: 'build' | 'walk'
) {
  if (stage === 'build') {
    if (isDevMode()) {
      if (hook) {
        ;(hook as any)(params)
      } else {
        const formattedMessage = message
          .split('\n')
          .filter((m) => m.trim().length > 0)
          .map((m) => `${colors.yellow('Warning:')} ${m}`)
          .join('\n')
        console.log(formattedMessage + '\n')
      }
    } else {
      throw new Error(message)
    }
  }
}
