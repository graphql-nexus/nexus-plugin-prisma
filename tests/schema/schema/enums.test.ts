import { enumType, inputObjectType, objectType } from '@nexus/schema'
import { generateSchemaAndTypes } from '../__utils'

it('publishes enum even as output type', async () => {
  const datamodel = `
datasource db {
  provider = "postgresql"
  url      = "postgresql://"
}

model User {
  id  Int @id @default(autoincrement())
  favouriteColor  Color
}

enum Color {
  Red
  Green
  Blue
}
  `

  const User = objectType({
    name: 'User',
    definition(t: any) {
      t.model.id()
      t.model.favouriteColor()
    },
  })

  const { schemaString: schema, typegen } = await generateSchemaAndTypes(datamodel, [User])

  expect(schema).toMatchSnapshot('schema')
  expect(typegen).toMatchSnapshot('typegen')
})

it('does not publish enum twice (from input/output type)', async () => {
  const datamodel = `
  datasource db {
    provider = "postgresql"
    url      = "postgresql://"
  }

  model User {
    id  Int @id @default(autoincrement())
    favouriteColor  Color
  }

  enum Color {
    Red
    Green
    Blue
  }
`

  const User = objectType({
    name: 'User',
    definition(t: any) {
      t.model.id()
      t.model.favouriteColor()
    },
  })

  const Query = objectType({
    name: 'Query',
    definition(t: any) {
      t.crud.findManyUser({ filtering: true })
    },
  })

  const { schemaString: schema, typegen } = await generateSchemaAndTypes(datamodel, [Query, User])

  expect(schema).toMatchSnapshot('schema')
  expect(typegen).toMatchSnapshot('typegen')
})

it('does not automatically publish input/enum type if already created by user', async () => {
  const datamodel = `
  datasource db {
    provider = "postgresql"
    url      = "postgresql://"
  }

  model User {
    id  Int @id @default(autoincrement())
    favouriteColor  Color
  }

  enum Color {
    Red
    Green
    Blue
  }
`

  const User = objectType({
    name: 'User',
    definition(t: any) {
      t.model.id()
      t.model.favouriteColor()
    },
  })

  const Color = enumType({
    name: 'Color',
    members: ['Something', 'Else'],
  })

  const Query = objectType({
    name: 'Query',
    definition(t: any) {
      t.crud.findManyUser({ filtering: true })
    },
  })

  const UserWhereInput = inputObjectType({
    name: 'UserWhereInput',
    definition(t) {
      t.string('customField')
    },
  })

  const { schemaString: schema, typegen } = await generateSchemaAndTypes(datamodel, [
    Query,
    User,
    Color,
    UserWhereInput,
  ])

  expect(schema).toMatchSnapshot('schema')
  expect(typegen).toMatchSnapshot('typegen')
})
