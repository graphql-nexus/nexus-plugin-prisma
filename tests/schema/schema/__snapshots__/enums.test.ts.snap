// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`does not automatically publish input/enum type if already created by user: schema 1`] = `
"enum Color {
  Something
  Else
}

type Query {
  findManyUser(where: UserWhereInput, first: Int, last: Int, before: UserWhereUniqueInput, after: UserWhereUniqueInput): [User!]!
}

type User {
  id: Int!
  favouriteColor: Color!
}

input UserWhereInput {
  customField: String
}

input UserWhereUniqueInput {
  id: Int
}
"
`;

exports[`does not automatically publish input/enum type if already created by user: typegen 1`] = `
"import * as Typegen from 'nexus-plugin-prisma/typegen'
import * as Prisma from '@prisma/client';

// Pagination type
type Pagination = {
  first?: boolean
  last?: boolean
  before?: boolean
  after?: boolean
}

// Prisma custom scalar names
type CustomScalars = 'No custom scalars are used in your Prisma Schema.'

// Prisma model type definitions
interface PrismaModels {
  User: Prisma.User
}

// Prisma input types metadata
interface NexusPrismaInputs {
  Query: {
    findManyUser: {
      filtering: 'AND' | 'OR' | 'NOT' | 'id' | 'favouriteColor'
      ordering: 'id' | 'favouriteColor'
    }
  },
  User: {

  }
}

// Prisma output types metadata
interface NexusPrismaOutputs {
  Query: {
    findOneUser: 'User'
    findManyUser: 'User'
  },
  Mutation: {
    createOneUser: 'User'
    updateOneUser: 'User'
    updateManyUser: 'BatchPayload'
    deleteOneUser: 'User'
    deleteManyUser: 'BatchPayload'
    upsertOneUser: 'User'
  },
  User: {
    id: 'Int'
    favouriteColor: 'Color'
  }
}

// Helper to gather all methods relative to a model
interface NexusPrismaMethods {
  User: Typegen.NexusPrismaFields<'User'>
  Query: Typegen.NexusPrismaFields<'Query'>
  Mutation: Typegen.NexusPrismaFields<'Mutation'>
}

interface NexusPrismaGenTypes {
  inputs: NexusPrismaInputs
  outputs: NexusPrismaOutputs
  methods: NexusPrismaMethods
  models: PrismaModels
  pagination: Pagination
  scalars: CustomScalars
}

declare global {
  interface NexusPrismaGen extends NexusPrismaGenTypes {}

  type NexusPrisma<
    TypeName extends string,
    ModelOrCrud extends 'model' | 'crud'
  > = Typegen.GetNexusPrisma<TypeName, ModelOrCrud>;
}
  "
`;

exports[`does not publish enum twice (from input/output type): schema 1`] = `
"enum Color {
  Red
  Green
  Blue
}

input IntFilter {
  equals: Int
  in: [Int!]
  notIn: [Int!]
  lt: Int
  lte: Int
  gt: Int
  gte: Int
  not: NestedIntFilter
}

input NestedIntFilter {
  equals: Int
  in: [Int!]
  notIn: [Int!]
  lt: Int
  lte: Int
  gt: Int
  gte: Int
  not: NestedIntFilter
}

type Query {
  findManyUser(where: UserWhereInput, first: Int, last: Int, before: UserWhereUniqueInput, after: UserWhereUniqueInput): [User!]!
}

type User {
  id: Int!
  favouriteColor: Color!
}

input UserWhereInput {
  AND: [UserWhereInput!]
  OR: [UserWhereInput!]
  NOT: [UserWhereInput!]
  id: IntFilter
  favouriteColor: Color
}

input UserWhereUniqueInput {
  id: Int
}
"
`;

exports[`does not publish enum twice (from input/output type): typegen 1`] = `
"import * as Typegen from 'nexus-plugin-prisma/typegen'
import * as Prisma from '@prisma/client';

// Pagination type
type Pagination = {
  first?: boolean
  last?: boolean
  before?: boolean
  after?: boolean
}

// Prisma custom scalar names
type CustomScalars = 'No custom scalars are used in your Prisma Schema.'

// Prisma model type definitions
interface PrismaModels {
  User: Prisma.User
}

// Prisma input types metadata
interface NexusPrismaInputs {
  Query: {
    findManyUser: {
      filtering: 'AND' | 'OR' | 'NOT' | 'id' | 'favouriteColor'
      ordering: 'id' | 'favouriteColor'
    }
  },
  User: {

  }
}

// Prisma output types metadata
interface NexusPrismaOutputs {
  Query: {
    findOneUser: 'User'
    findManyUser: 'User'
  },
  Mutation: {
    createOneUser: 'User'
    updateOneUser: 'User'
    updateManyUser: 'BatchPayload'
    deleteOneUser: 'User'
    deleteManyUser: 'BatchPayload'
    upsertOneUser: 'User'
  },
  User: {
    id: 'Int'
    favouriteColor: 'Color'
  }
}

// Helper to gather all methods relative to a model
interface NexusPrismaMethods {
  User: Typegen.NexusPrismaFields<'User'>
  Query: Typegen.NexusPrismaFields<'Query'>
  Mutation: Typegen.NexusPrismaFields<'Mutation'>
}

interface NexusPrismaGenTypes {
  inputs: NexusPrismaInputs
  outputs: NexusPrismaOutputs
  methods: NexusPrismaMethods
  models: PrismaModels
  pagination: Pagination
  scalars: CustomScalars
}

declare global {
  interface NexusPrismaGen extends NexusPrismaGenTypes {}

  type NexusPrisma<
    TypeName extends string,
    ModelOrCrud extends 'model' | 'crud'
  > = Typegen.GetNexusPrisma<TypeName, ModelOrCrud>;
}
  "
`;

exports[`publishes enum even as output type: schema 1`] = `
"enum Color {
  Red
  Green
  Blue
}

type Query {
  ok: Boolean!
}

type User {
  id: Int!
  favouriteColor: Color!
}
"
`;

exports[`publishes enum even as output type: typegen 1`] = `
"import * as Typegen from 'nexus-plugin-prisma/typegen'
import * as Prisma from '@prisma/client';

// Pagination type
type Pagination = {
  first?: boolean
  last?: boolean
  before?: boolean
  after?: boolean
}

// Prisma custom scalar names
type CustomScalars = 'No custom scalars are used in your Prisma Schema.'

// Prisma model type definitions
interface PrismaModels {
  User: Prisma.User
}

// Prisma input types metadata
interface NexusPrismaInputs {
  Query: {
    findManyUser: {
      filtering: 'AND' | 'OR' | 'NOT' | 'id' | 'favouriteColor'
      ordering: 'id' | 'favouriteColor'
    }
  },
  User: {

  }
}

// Prisma output types metadata
interface NexusPrismaOutputs {
  Query: {
    findOneUser: 'User'
    findManyUser: 'User'
  },
  Mutation: {
    createOneUser: 'User'
    updateOneUser: 'User'
    updateManyUser: 'BatchPayload'
    deleteOneUser: 'User'
    deleteManyUser: 'BatchPayload'
    upsertOneUser: 'User'
  },
  User: {
    id: 'Int'
    favouriteColor: 'Color'
  }
}

// Helper to gather all methods relative to a model
interface NexusPrismaMethods {
  User: Typegen.NexusPrismaFields<'User'>
  Query: Typegen.NexusPrismaFields<'Query'>
  Mutation: Typegen.NexusPrismaFields<'Mutation'>
}

interface NexusPrismaGenTypes {
  inputs: NexusPrismaInputs
  outputs: NexusPrismaOutputs
  methods: NexusPrismaMethods
  models: PrismaModels
  pagination: Pagination
  scalars: CustomScalars
}

declare global {
  interface NexusPrismaGen extends NexusPrismaGenTypes {}

  type NexusPrisma<
    TypeName extends string,
    ModelOrCrud extends 'model' | 'crud'
  > = Typegen.GetNexusPrisma<TypeName, ModelOrCrud>;
}
  "
`;
